### ISSUE TITLE
Quirky sort behavior with #dbrefs and setunnion()

### ISSUE THREAD

**Issue by Chalice** [pennmush/pennmush#1274]
Greetings,

There seems to be rather quirky behavior when it comes to setunion (and probably setinter/setdiff as well) and #dbrefs.

The output of 'setunion(#3 #4 #5, #453,  ,n)' simply returns '#3', while I'd expect it to be at least include the full set of #dbrefs, even when the sorttype might not be able to correctly sort them. Tested on the latest Penn GIT release, as well as M*U*S*H,

Regards,
--Ambrosia@RhostMUSH

**Comment by grapenut** [comment]
Passing a sort type will attempt to cast the list items into that type. When you cast a string that doesn't start with a numeral into a number, its value is 0. The internal representation for all those dbrefs is a 0, so extra 0s get filtered out and only the first element is returned.

**Comment by shawnw** [comment]
I'm not sure about always notifying in this case... hmm. Maybe extending the warning system to softcode things might be a possibility? Or just a flag that turns on some extra diagnostics and checks?

**Comment by talvo** [comment]
I'm not sure I agree with a notification for this; the entire point of specifying a sort type is to force something that may not match the default. Spamming people to tell them the function will work specifically as they've requested seems annoying in typical circumstances...

**Comment by grapenut** [comment]
Having the DEBUG flag set while simultaneously trying to debug this particular function where the given sort type is different from the auto-detected sort type is not a typical circumstance. I don't think a single line of warning embedded in the DEBUG output would be anything close to as "spammy" as the rest of the DEBUG output.

I don't care either way. But this is not the first time this issue has been raised as a bug, and debugging it requires experts looking at the hardcode.

**Comment by talvo** [comment]
If you're going to make it DEBUG-only - I'd rather go further and have the debug actually show the internal representation generated for each item.

**Comment by shawnw** [comment]
I think I'm just going to add a note to HELP SORTING about unexpected things happening when the declared sort type doesn't match the provided data, and close this.

**Comment by grapenut** [comment]
@shawnw I agree, this is probably the least invasive solution. I would suggest providing a specific warning of the most likely pitfall, e.g. alpha strings being converted to 0 during a numerical sort.

Perhaps something like:
> Passing an explicit sort type will attempt to cast the list items into that type. A note of warning, when you cast a string that doesn't start with an integer into a numerical type, its value will default to 0. For example, when using a numerical sort type the string 'abc123' will become 0 while '123abc' would become 123.
